src/persist/README.md

```diff
@@ -3,49 +3,194 @@
-given timestamp. It could implemented on top of a SQLite table with something
-like the following schema:
+A persistence user initializes a `Buffer` and a `Blob` implementation and uses
+them to start the runtime, receiving a `RuntimeClient` in response. This client
+is the primary external interface to the persistence system. This runtime allows
+us to funnel everything a process is persisting through a single place for
+batching and rate limiting. The buffer and blob are exclusively locked for
+writes by this process (they can be opened for reads by other processes,
+however).
 
-[persister v1]: https://github.com/danhhz/differential-dataflow/blob/02673114b05933341893ab603327237a9583e432/persist/src/persister.rs#L30-L39
+The runtime stops when every copy of the client has been dropped (or when any
+client's stop method has been called), attempting to first gracefully finish any
+pending work, but rejecting new work.
 
-```sql
-CREATE TABLE persisted (persisted_id INT, key BYTES, val BYTES, ts BYTES, diff BYTES);
-CREATE INDEX ON persisted (persisted_id, ts, key);
+As needed, the user creates a persisted stream by name, returning a token. At
+dataflow construction time, this token can be exchanged to create one of several
+available persistence operators (see below). When the operator is run, it hands
+any blocking io or cpu heavy work to the runtime, asking the dataflow to be
+scheduled again when this finishes. If the same buffer and blob implementations
+were previously used to persist a stream with the same name, this previous data
+is read and also output by the operator, yielding periodically to allow
+downstream operators a chance to process it.
+
+## Operator Guarantees
+
+- _idempotence_: Any record emitted by a persistence operator is guaranteed to
+  be included in the output of all future operators for the same persistent
+  stream. This is true even before the record's timestamp has been closed.
+- _ordering_: The same guarantees you'd expect out of any differential
+  collection or timely stream. TODO: Is there anything more specific to say
+  here?
+- _determinism_: TODO: This is easy in the absence of errors from the storage,
+  but that's not the interesting part. "S3 is down" is obviously a transient
+  error, but even things like permissions errors are transient in terms of the
+  strict definition of deterministic we're using. I think the Materialize "error
+  stream" model is potentially a good way to model this where anything on the
+  main output is deterministic at some timestamp iff the error stream is an
+  empty collection at that timestamp. Think about this more.
+
+## Example: Persisting a Stream
+
+NB: This operator doesn't exist yet, but this is more or less what the interface
+will look like.
+
+```rust
+let persister: RuntimeClient = ...
+let token = persister.create_or_load("example stream")?;
+
+let stream: Stream<(K, V, T, R)> = ...
+let (persisted_stream, err_stream) = stream.new_persisted_stream(token);
```

Comment:
Internal to `dataflow`, the streams are logically equivalent to `Stream<_, (Result<(K, V), Error>, Time, Diff)>`, if that helps. We use two separate streams a lot because the operators get annoyed by the continual demultiplexing, but if it makes your life easier to think of it as a single collection, feel free.

Reply:
It mainly seems to make it easier to fail to forget about error streams, which .. elsewhere in the codebase we manage to do when one can just accept a stream of `Row` or what have you. No strong feels, but I didn't want you to accidentally import an idiom (two streams) that exists mostly for interop with DD operators.

Reply:
Removed everything about errors in the new version until someone gets a chance to figure out how they should work.